xarray/test/test_quantities.py

```diff
@@ -0,0 +1,64 @@
+import numpy as np
+import pandas as pd
+
+from xarray import (align, broadcast, Dataset, DataArray, Variable)
+
+from xarray.test import (TestCase, unittest)
+
+try:
+    import quantities as pq
+    has_quantities = True
+except ImportError:
+    has_quantities = False
+
+def requires_quantities(test):
+    return test if has_quantities else unittest.skip('requires python-quantities')(test)
+
+@requires_quantities
+class TestWithQuantities(TestCase):
+    def setUp(self):
+        self.x = np.arange(10) * pq.A
+        self.y = np.arange(20)
+        self.xp = np.arange(10) * pq.J
+        self.v = np.arange(10*20).reshape(10,20) * pq.V
+        self.da = DataArray(self.v,dims=['x','y'],coords=dict(x=self.x,y=self.y,xp=(['x'],self.xp)))
+
+    def assertEqualWUnits(self,a,b):
+        self.assertIsNotNone(getattr(a, 'units', None))
+        self.assertIsNotNone(getattr(b, 'units', None))
+        self.assertEqual(a.units,b.units)
+        np.testing.assert_allclose(a.magnitude,b.magnitude)
+
+    def test_units_in_data_and_coords(self):
+        da = self.da
+        self.assertEqualWUnits(da.xp.data,self.xp)
+        self.assertEqualWUnits(da.data,self.v)
+
+    def test_arithmetics(self):
+        da = self.da
+        f = DataArray(np.arange(10*20).reshape(10,20)*pq.A,dims=['x','y'],coords=dict(x=self.x,y=self.y))
+        self.assertEqualWUnits((da*f).data, da.data*f.data)
```

Comment:
What about arithmetic that requires broadcasting? e.g., between arrays with different dimensions and/or different dimension order?

That will require ensuring that `ops.transpose` and `ops.broadcast_to` work on subclasses (e.g., `xarray.Variable.transpose` and `xarray.Variable.expand_dims`). I think the former *should* work already because of NumPy, but `broadcast_to` will probably require adding `subok=True`.